In [1]:
#lets getting dataset from the kaggle

import pandas as pd
import numpy as np
import statistics
import numpy as np
    
df = pd.read_csv('/kaggle/input/palmer-archipelago-penguin-data/penguins_lter.csv')

In [2]:
df.shape

(344, 17)

In [3]:
drop_feature = ['studyName','Sample Number','Date Egg','Comments','Date Egg']
# Drop the specified columns
df = df.drop(columns=drop_feature)
df.head(10)

Species  Region     Island               Stage  \
0  Adelie Penguin (Pygoscelis adeliae)  Anvers  Torgersen  Adult, 1 Egg Stage   
1  Adelie Penguin (Pygoscelis adeliae)  Anvers  Torgersen  Adult, 1 Egg Stage   
2  Adelie Penguin (Pygoscelis adeliae)  Anvers  Torgersen  Adult, 1 Egg Stage   
3  Adelie Penguin (Pygoscelis adeliae)  Anvers  Torgersen  Adult, 1 Egg Stage   
4  Adelie Penguin (Pygoscelis adeliae)  Anvers  Torgersen  Adult, 1 Egg Stage   
5  Adelie Penguin (Pygoscelis adeliae)  Anvers  Torgersen  Adult, 1 Egg Stage   
6  Adelie Penguin (Pygoscelis adeliae)  Anvers  Torgersen  Adult, 1 Egg Stage   
7  Adelie Penguin (Pygoscelis adeliae)  Anvers  Torgersen  Adult, 1 Egg Stage   
8  Adelie Penguin (Pygoscelis adeliae)  Anvers  Torgersen  Adult, 1 Egg Stage   
9  Adelie Penguin (Pygoscelis adeliae)  Anvers  Torgersen  Adult, 1 Egg Stage   

  Individual ID Clutch Completion  Culmen Length (mm)  Culmen Depth (mm)  \
0          N1A1               Yes                39.1               18.7   
1          N1A2               Yes                39.5               17.4   
2          N2A1               Yes                40.3               18.0   
3          N2A2               Yes                 NaN                NaN   
4          N3A1               Yes                36.7               19.3   
5          N3A2               Yes                39.3               20.6   
6          N4A1                No                38.9               17.8   
7          N4A2                No                39.2               19.6   
8          N5A1               Yes                34.1               18.1   
9          N5A2               Yes                42.0               20.2   

   Flipper Length (mm)  Body Mass (g)     Sex  Delta 15 N (o/oo)  \
0                181.0         3750.0    MALE                NaN   
1                186.0         3800.0  FEMALE            8.94956   
2                195.0         3250.0  FEMALE            8.36821   
3                  NaN            NaN     NaN                NaN   
4                193.0         3450.0  FEMALE            8.76651   
5                190.0         3650.0    MALE            8.66496   
6                181.0         3625.0  FEMALE            9.18718   
7                195.0         4675.0    MALE            9.46060   
8                193.0         3475.0     NaN                NaN   
9                190.0         4250.0     NaN            9.13362   

   Delta 13 C (o/oo)  
0                NaN  
1          -24.69454  
2          -25.33302  
3                NaN  
4          -25.32426  
5          -25.29805  
6          -25.21799  
7          -24.89958  
8                NaN  
9          -25.09368

In [4]:
#this def helps out changing NaN to Mean Value of column that are being worked on.

def nan_to_mean(mylist):
    non_nan = list(filter(lambda x: not np.isnan(x),mylist))
    mean_value = round(np.mean(non_nan),3)
    
    filled_list = [mean_value if np.isnan(x) else x for x in mylist]
    return filled_list
# it is returning the list that fitted with mean value instead of NaN

In [5]:
#this def helps out that if encounters non_sense value such as '.' or '+' ,'_' etc, it will fill with 'NaN' value so that it will be completed 
#it with mean value so that the weight of the feeture is not disrupted.
def char_to_nan(mylist):
    temp_list= df['Sex']
    my_nan = temp_list[3]
    temp_list = [my_nan if item == '.' else item for item in temp_list]

    for each in range(len(temp_list)):
        if temp_list[each] == '.':
            print(each)
    mylist = temp_list
    return mylist

In [6]:
#let see which column/feature has missing values and how much? According to that info, we need missing features so that it goes to nan_to_mean methods.
def missing_value (mylist):
    missing_values = mylist.isna().sum()
    nan_list =[]
    print("Missing values in each column:")
    for k,v in missing_values.items():
        print(f"{k}: {v / df.shape[0] * 100:.2f}%")
        if v != 0:
            nan_list += [k]
    return nan_list

In [7]:
#here, what i aimed is to see what kind of outputs/varients are there in categorical features: after that, i need to make them as numerical values
#so i can get them into training process of my model.

non_numerical = ['Species','Region','Island','Stage','Individual ID','Sex']

def varienty_search(feature,mylist):
    mycheck = []
    for each in mylist:
        if each not in mycheck:
            mycheck += [each] 
    print(feature,':',mycheck)


for each in (df[non_numerical]):
    varienty_search(each, df[each])

Species : ['Adelie Penguin (Pygoscelis adeliae)', 'Chinstrap penguin (Pygoscelis antarctica)', 'Gentoo penguin (Pygoscelis papua)']
Region : ['Anvers']
Island : ['Torgersen', 'Biscoe', 'Dream']
Stage : ['Adult, 1 Egg Stage']
Individual ID : ['N1A1', 'N1A2', 'N2A1', 'N2A2', 'N3A1', 'N3A2', 'N4A1', 'N4A2', 'N5A1', 'N5A2', 'N6A1', 'N6A2', 'N7A1', 'N7A2', 'N8A1', 'N8A2', 'N9A1', 'N9A2', 'N10A1', 'N10A2', 'N11A1', 'N11A2', 'N12A1', 'N12A2', 'N13A1', 'N13A2', 'N17A1', 'N17A2', 'N18A1', 'N18A2', 'N21A1', 'N21A2', 'N22A1', 'N22A2', 'N23A1', 'N23A2', 'N24A1', 'N24A2', 'N25A1', 'N25A2', 'N26A1', 'N26A2', 'N27A1', 'N27A2', 'N28A1', 'N28A2', 'N29A1', 'N29A2', 'N30A1', 'N30A2', 'N32A1', 'N32A2', 'N34A1', 'N34A2', 'N35A1', 'N35A2', 'N36A1', 'N36A2', 'N37A1', 'N37A2', 'N38A1', 'N38A2', 'N39A1', 'N39A2', 'N40A1', 'N40A2', 'N41A1', 'N41A2', 'N42A1', 'N42A2', 'N44A1', 'N44A2', 'N45A1', 'N45A2', 'N46A1', 'N46A2', 'N48A1', 'N48A2', 'N49A1', 'N49A2', 'N50A1', 'N50A2', 'N47A1', 'N47A2', 'N51A1', 'N51A2', 'N

In [8]:
#preprocessing for non_numerical features. on the other words, it makes them as categorical
def preprocessing(df):
    Species_class = {'Adelie Penguin (Pygoscelis adeliae)': 0,
                 'Chinstrap penguin (Pygoscelis antarctica)':1, 
                 'Gentoo penguin (Pygoscelis papua)':2}
    df['Species']= [Species_class[each] for each in df['Species']]

    
    Region_class = {'Anvers':0}
    df['Region']= [Region_class[each] for each in df['Region']]

    Island_class = {'Torgersen':0,
                'Biscoe':1, 
                'Dream':2}
    df['Island']= [Island_class[each] for each in df['Island']]

    Stage_class = {'Adult, 1 Egg Stage':0}
    df['Stage']= [Stage_class[each] for each in df['Stage']]

    Clutch_class = {'Yes':0, 'No':1}
    df['Clutch Completion']= [Clutch_class[each] for each in df['Clutch Completion']]

    counter = 0
    Individual_ID = []
    for each in range(len(df['Individual ID'])):
        Individual_ID += [counter] 
        counter+=1
    df['Individual ID'] = Individual_ID
    
    dp = df['Sex']
    for counter in range(len(dp)):
        each = dp.loc[counter]
        if each == 'FEMALE':
            dp.at[counter] = 0
        elif each == 'MALE':
            dp.at[counter] = 1
    df['Sex'] = dp

In [9]:
preprocessing(df)

In [10]:
df.head(10)

Species  Region  Island  Stage  Individual ID  Clutch Completion  \
0        0       0       0      0              0                  0   
1        0       0       0      0              1                  0   
2        0       0       0      0              2                  0   
3        0       0       0      0              3                  0   
4        0       0       0      0              4                  0   
5        0       0       0      0              5                  0   
6        0       0       0      0              6                  1   
7        0       0       0      0              7                  1   
8        0       0       0      0              8                  0   
9        0       0       0      0              9                  0   

   Culmen Length (mm)  Culmen Depth (mm)  Flipper Length (mm)  Body Mass (g)  \
0                39.1               18.7                181.0         3750.0   
1                39.5               17.4                186.0         3800.0   
2                40.3               18.0                195.0         3250.0   
3                 NaN                NaN                  NaN            NaN   
4                36.7               19.3                193.0         3450.0   
5                39.3               20.6                190.0         3650.0   
6                38.9               17.8                181.0         3625.0   
7                39.2               19.6                195.0         4675.0   
8                34.1               18.1                193.0         3475.0   
9                42.0               20.2                190.0         4250.0   

   Sex  Delta 15 N (o/oo)  Delta 13 C (o/oo)  
0    1                NaN                NaN  
1    0            8.94956          -24.69454  
2    0            8.36821          -25.33302  
3  NaN                NaN                NaN  
4    0            8.76651          -25.32426  
5    1            8.66496          -25.29805  
6    0            9.18718          -25.21799  
7    1            9.46060          -24.89958  
8  NaN                NaN                NaN  
9  NaN            9.13362          -25.09368

In [11]:
'''
Species : ['Adelie Penguin (Pygoscelis adeliae)', 'Chinstrap penguin (Pygoscelis antarctica)', 'Gentoo penguin (Pygoscelis papua)']
Region : ['Anvers']
Island : ['Torgersen', 'Biscoe', 'Dream']
Stage : ['Adult, 1 Egg Stage']
Individual ID : ['N1A1', 'N1A2', 'N2A1', 'N2A2', 'N3A1', 'N3A2', 'N4A1', 'N4A2', 'N5A1', 'N5A2', 'N6A1', 'N6A2', 'N7A1', 'N7A2', 'N8A1', 'N8A2', 'N9A1', 'N9A2', 'N10A1', 'N10A2', 'N11A1', 'N11A2', 'N12A1', 'N12A2', 'N13A1', 'N13A2', 'N17A1', 'N17A2', 'N18A1', 'N18A2', 'N21A1', 'N21A2', 'N22A1', 'N22A2', 'N23A1', 'N23A2', 'N24A1', 'N24A2', 'N25A1', 'N25A2', 'N26A1', 'N26A2', 'N27A1', 'N27A2', 'N28A1', 'N28A2', 'N29A1', 'N29A2', 'N30A1', 'N30A2', 'N32A1', 'N32A2', 'N34A1', 'N34A2', 'N35A1', 'N35A2', 'N36A1', 'N36A2', 'N37A1', 'N37A2', 'N38A1', 'N38A2', 'N39A1', 'N39A2', 'N40A1', 'N40A2', 'N41A1', 'N41A2', 'N42A1', 'N42A2', 'N44A1', 'N44A2', 'N45A1', 'N45A2', 'N46A1', 'N46A2', 'N48A1', 'N48A2', 'N49A1', 'N49A2', 'N50A1', 'N50A2', 'N47A1', 'N47A2', 'N51A1', 'N51A2', 'N53A1', 'N53A2', 'N55A1', 'N55A2', 'N58A1', 'N58A2', 'N60A1', 'N60A2', 'N61A1', 'N61A2', 'N63A1', 'N63A2', 'N64A1', 'N64A2', 'N66A1', 'N66A2', 'N67A1', 'N67A2', 'N69A1', 'N69A2', 'N71A1', 'N71A2', 'N72A1', 'N72A2', 'N73A1', 'N73A2', 'N76A1', 'N76A2', 'N77A1', 'N77A2', 'N78A1', 'N78A2', 'N79A1', 'N79A2', 'N80A1', 'N80A2', 'N81A1', 'N81A2', 'N82A1', 'N82A2', 'N83A1', 'N83A2', 'N84A1', 'N84A2', 'N85A1', 'N85A2', 'N62A1', 'N62A2', 'N68A1', 'N68A2', 'N70A1', 'N70A2', 'N89A1', 'N89A2', 'N65A1', 'N65A2', 'N74A1', 'N74A2', 'N75A1', 'N75A2', 'N86A1', 'N86A2', 'N87A1', 'N87A2', 'N88A1', 'N88A2', 'N90A1', 'N90A2', 'N92A1', 'N92A2', 'N93A1', 'N93A2', 'N94A1', 'N94A2', 'N95A1', 'N95A2', 'N96A1', 'N96A2', 'N98A1', 'N98A2', 'N99A1', 'N99A2', 'N100A1', 'N100A2', 'N31A1', 'N31A2', 'N33A1', 'N33A2', 'N56A1', 'N56A2', 'N14A1', 'N14A2', 'N15A1', 'N15A2', 'N16A1', 'N16A2', 'N19A1', 'N19A2', 'N20A1', 'N20A2', 'N54A1', 'N54A2', 'N43A1', 'N43A2']
Sex : ['MALE', 'FEMALE', nan, '.']
'''
# we observed that undefined output is there in sex, so we need to change it with nan and then with mean value 
df.Sex = char_to_nan(df.Sex)
df.Sex = nan_to_mean(df.Sex)


In [12]:
#It works properly. This result before char_to_nan and nan_to_mean has '.' and 'nan' values. right now we got rid of them by adding mean values.
varienty_search('Sex',df.Sex)

Sex : [1.0, 0.0, 0.505]


In [13]:
#after replacing '.' with non then with mean values, there wont be any undefined characters in my data.
k=missing_value (df)
print('Nan columns are: \n',k)

Missing values in each column:
Species: 0.00%
Region: 0.00%
Island: 0.00%
Stage: 0.00%
Individual ID: 0.00%
Clutch Completion: 0.00%
Culmen Length (mm): 0.58%
Culmen Depth (mm): 0.58%
Flipper Length (mm): 0.58%
Body Mass (g): 0.58%
Sex: 0.00%
Delta 15 N (o/oo): 4.07%
Delta 13 C (o/oo): 3.78%
Nan columns are: 
 ['Culmen Length (mm)', 'Culmen Depth (mm)', 'Flipper Length (mm)', 'Body Mass (g)', 'Delta 15 N (o/oo)', 'Delta 13 C (o/oo)']


In [14]:
#So it is time to fill missing values with means
for each in k:
    df[each]=nan_to_mean(df[each])

In [15]:
df.head(10)

Species  Region  Island  Stage  Individual ID  Clutch Completion  \
0        0       0       0      0              0                  0   
1        0       0       0      0              1                  0   
2        0       0       0      0              2                  0   
3        0       0       0      0              3                  0   
4        0       0       0      0              4                  0   
5        0       0       0      0              5                  0   
6        0       0       0      0              6                  1   
7        0       0       0      0              7                  1   
8        0       0       0      0              8                  0   
9        0       0       0      0              9                  0   

   Culmen Length (mm)  Culmen Depth (mm)  Flipper Length (mm)  Body Mass (g)  \
0              39.100             18.700              181.000       3750.000   
1              39.500             17.400              186.000       3800.000   
2              40.300             18.000              195.000       3250.000   
3              43.922             17.151              200.915       4201.754   
4              36.700             19.300              193.000       3450.000   
5              39.300             20.600              190.000       3650.000   
6              38.900             17.800              181.000       3625.000   
7              39.200             19.600              195.000       4675.000   
8              34.100             18.100              193.000       3475.000   
9              42.000             20.200              190.000       4250.000   

     Sex  Delta 15 N (o/oo)  Delta 13 C (o/oo)  
0  1.000            8.73300          -25.68600  
1  0.000            8.94956          -24.69454  
2  0.000            8.36821          -25.33302  
3  0.505            8.73300          -25.68600  
4  0.000            8.76651          -25.32426  
5  1.000            8.66496          -25.29805  
6  0.000            9.18718          -25.21799  
7  1.000            9.46060          -24.89958  
8  0.505            8.73300          -25.68600  
9  0.505            9.13362          -25.09368

In [16]:
missing_value(df)

# now we will see that there is no missing value in each column

Missing values in each column:
Species: 0.00%
Region: 0.00%
Island: 0.00%
Stage: 0.00%
Individual ID: 0.00%
Clutch Completion: 0.00%
Culmen Length (mm): 0.00%
Culmen Depth (mm): 0.00%
Flipper Length (mm): 0.00%
Body Mass (g): 0.00%
Sex: 0.00%
Delta 15 N (o/oo): 0.00%
Delta 13 C (o/oo): 0.00%


[]

In [17]:
'''
here, all data is ready to train right now because there is no non_sense value stored in each column
'''

'\nhere, all data is ready to train right now because there is no non_sense value stored in each column\n'

In [18]:
#lets go with StandardScaler in scaling

from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()
scaled_df = scaler.fit_transform(df)
scaled_df = pd.DataFrame(scaled_df, columns=df.columns)


In [19]:
scaled_df.head(10)

#the reason why we used standard scaler is that each value in each column has different weight. In order to distribute the weights and prevent from manipulating by high values, we need to use it. 

Species  Region    Island  Stage  Individual ID  Clutch Completion  \
0 -1.029802     0.0 -1.768053    0.0      -1.727023          -0.341882   
1 -1.029802     0.0 -1.768053    0.0      -1.716953          -0.341882   
2 -1.029802     0.0 -1.768053    0.0      -1.706883          -0.341882   
3 -1.029802     0.0 -1.768053    0.0      -1.696813          -0.341882   
4 -1.029802     0.0 -1.768053    0.0      -1.686743          -0.341882   
5 -1.029802     0.0 -1.768053    0.0      -1.676673          -0.341882   
6 -1.029802     0.0 -1.768053    0.0      -1.666602           2.924988   
7 -1.029802     0.0 -1.768053    0.0      -1.656532           2.924988   
8 -1.029802     0.0 -1.768053    0.0      -1.646462          -0.341882   
9 -1.029802     0.0 -1.768053    0.0      -1.636392          -0.341882   

   Culmen Length (mm)  Culmen Depth (mm)  Flipper Length (mm)  Body Mass (g)  \
0           -0.887081           0.787743            -1.422488  -5.657892e-01   
1           -0.813494           0.126557            -1.065352  -5.031679e-01   
2           -0.666320           0.431720            -0.422507  -1.192003e+00   
3            0.000013          -0.000086            -0.000015  -4.805823e-07   
4           -1.328605           1.092906            -0.565361  -9.415172e-01   
5           -0.850288           1.754092            -0.779643  -6.910319e-01   
6           -0.923875           0.329999            -1.422488  -7.223425e-01   
7           -0.868684           1.245487            -0.422507   5.927054e-01   
8           -1.806922           0.482580            -0.565361  -9.102065e-01   
9           -0.353574           1.550650            -0.779643   6.042409e-02   

        Sex  Delta 15 N (o/oo)  Delta 13 C (o/oo)  
0  1.007234          -0.000679           0.000361  
1 -1.025613           0.400651           1.275325  
2 -1.025613          -0.676708           0.454274  
3  0.000975          -0.000679           0.000361  
4 -1.025613           0.061422           0.465539  
5  1.007234          -0.126770           0.499244  
6 -1.025613           0.841009           0.602197  
7  1.007234           1.347712           1.011655  
8  0.000975          -0.000679           0.000361  
9  0.000975           0.741751           0.762053

In [20]:
'''
In machine learning, it is common practice to scale features so that they have a similar range of values. This helps to prevent features with larger ranges from dominating
those with smaller ranges, which can lead to biased or suboptimal models.
'''

from sklearn.preprocessing import MinMaxScaler, OneHotEncoder
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import cross_val_predict, KFold
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score


y = df.Species
features = ['Region', 'Island', 'Stage', 'Clutch Completion', 
            'Culmen Length (mm)', 'Culmen Depth (mm)', 'Flipper Length (mm)', 
            'Body Mass (g)', 'Sex', 'Delta 15 N (o/oo)', 'Delta 13 C (o/oo)']
X = scaled_df[features]

#Preprocessing pipeline

preprocessor = ColumnTransformer(
    transformers=[
        ('num', StandardScaler(), ['Culmen Length (mm)', 'Culmen Depth (mm)', 'Flipper Length (mm)', 'Body Mass (g)', 'Delta 15 N (o/oo)', 'Delta 13 C (o/oo)']),
        ('cat', OneHotEncoder(), ['Region', 'Island', 'Stage', 'Clutch Completion', 'Sex'])
    ])

# Create a pipeline with preprocessing and the classifier
model = Pipeline(steps=[
    ('preprocessor', preprocessor),
    ('classifier', RandomForestClassifier(random_state=42))])

# Set up K-Fold Cross-Validation
kfold = KFold(n_splits=5, random_state=42, shuffle=True)

# Perform cross-validation and get predictions
predictions = cross_val_predict(model, X, y, cv=kfold)

# Perform cross-validation and get accuracy scores
scores = cross_val_score(model, X, y, cv=kfold, scoring='accuracy')

# Output the results
print(f"Accuracy for each fold: {scores}")
print(f"Average Accuracy: {scores.mean()}")

# Output predictions
print("Predictions:")
print(predictions)

# Evaluate the predictions
print(f"Overall Accuracy: {accuracy_score(y, predictions)}")


Accuracy for each fold: [1.         0.98550725 0.98550725 1.         0.97058824]
Average Accuracy: 0.9883205456095482
Predictions:
[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 0 1 1 1 1 1 1 1 0 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 2 2
 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2
 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2
 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2
 2 2 2 2 2 2 0 2 2 2 2]
Overall Accuracy: 0.9883720930232558


In [21]:
mydict = {
    'pred_y':predictions,
    'actual': y
}

In [22]:
outputresult = pd.DataFrame(mydict)

In [23]:
outputresult.iloc[320:343]

pred_y  actual
320       2       2
321       2       2
322       2       2
323       2       2
324       2       2
325       2       2
326       2       2
327       2       2
328       2       2
329       2       2
330       2       2
331       2       2
332       2       2
333       2       2
334       2       2
335       2       2
336       2       2
337       2       2
338       2       2
339       0       2
340       2       2
341       2       2
342       2       2

In [24]:
list(X.columns)

['Region',
 'Island',
 'Stage',
 'Clutch Completion',
 'Culmen Length (mm)',
 'Culmen Depth (mm)',
 'Flipper Length (mm)',
 'Body Mass (g)',
 'Sex',
 'Delta 15 N (o/oo)',
 'Delta 13 C (o/oo)']

In [25]:
'''
lets see the feature importances:
When you use cross-validation, each fold involves fitting a new model on a subset of the data, so the model in the pipeline is not fitted on the entire dataset at once. 
Therefore, you cannot directly access feature_importances_ from the model object because the fitting process occurs separately for each fold. Now, i need to re-do fitting process in each fold
then make each ones average.
'''


#last important thing is that as using k-folds, importance will exist in each fold. so i need to average them.

def get_feature_importances_from_cv(model, X, y, cv):
    from sklearn.base import clone
    importance = []
    
# cv.split(X, y) provides the indices that will be assigned to the training and testing sets for each fold in the K-fold cross-validation process.
    for train_index, test_index in cv.split(X,y):
        X_train, X_test = X.iloc[train_index], X.iloc[test_index]
        y_train, y_test = y.iloc[train_index], y.iloc[test_index]
        # right now, it is splitted form. 
        
        #as this is recurring until the last fold, the model has a tricky issue. if I would not refresh model that already fitted at the last fold, the output would be biased because of the last process.!!!
        #in order to prevent from having biased output, we need to CLONE model so that it ensure us to use model as if it is used in first time.
        # biased output can be generated by overfitting or underfitting which is not desired in ML.
        
        cloned_model = clone(model)
        cloned_model.fit(X_train,y_train)
      
       

        
        #just to refresh my memory
        model = Pipeline(steps=[
            ('preprocessor', preprocessor),
            ('classifier', RandomForestClassifier(random_state=42))])
       
        
        rf_classifier = cloned_model.named_steps['classifier']  # This is how to reach out any elements in pipeline by just writing named_steps ['xxx']
        weights = rf_classifier.feature_importances_
        print(weights)
        importance.append(weights)    # it will be filled with new importances by each k-fold
        
    # Average the feature importances across all folds
    avg_importances = np.mean(importance, axis=0)
    
    return avg_importances


In [26]:
''' 
# Set up K-Fold Cross-Validation
kfold = KFold(n_splits=5, random_state=42, shuffle=True)
I declared it before like this above, lets call kfold back
'''
avg_importance = get_feature_importances_from_cv(model,X,y,kfold)


[0.22344571 0.16421295 0.20794362 0.08115346 0.05762412 0.14779713
 0.         0.01365484 0.05018366 0.0452972  0.         0.0011884
 0.00173625 0.00216851 0.00074252 0.00285163]
[0.23131357 0.13990359 0.18212138 0.09856631 0.05609786 0.18103997
 0.         0.01557887 0.053351   0.03544708 0.         0.00066525
 0.00118415 0.00242097 0.00041777 0.00189223]
[0.23636695 0.15779181 0.20291231 0.07932176 0.06562997 0.13904606
 0.         0.01724646 0.04580771 0.04716118 0.         0.00236062
 0.00180358 0.00253437 0.00054281 0.00147441]
[0.24492103 0.13420824 0.19716633 0.08243711 0.04921616 0.14548606
 0.         0.01972115 0.05805075 0.05954536 0.         0.00229216
 0.00226885 0.00159649 0.00079939 0.00229093]
[0.22345168 0.14089122 0.19700745 0.08036139 0.0542448  0.1752273
 0.         0.01629365 0.05365829 0.05224148 0.         0.00087104
 0.00099449 0.00150392 0.00163809 0.00161519]


### What I encountered the thing i found it very important is that:

when i wrote like this:
    for train_index, test_index in cv.split(X,y):
        X_train, X_test = X[train_index], X[test_index]
        y_train, y_test = y[train_index], y[test_index]
        # right now, it is splitted form. 
        
**I got a error**:

"None of [Index([  0,   1,   2,   4,   6,   7,   8,  10,  11,  12,\n       ...\n       331, 332, 334, 335, 337, 338, 339, 340, 342, 343],\n      dtype='int64', length=275)] are in the [columns]"

k-fold ensures the model to get train/test data by integer valued indices. Here, I need to use '.iloc'. '.iloc' serves to get positioned data. 

Here is the refined:
    for train_index, test_index in cv.split(X,y):
        X_train, X_test = X.iloc[train_index], X.iloc[test_index]
        y_train, y_test = y.iloc[train_index], y.iloc[test_index]
        
After that, train and test data will be formed from X and Y with specific position that defined in each k-fold.


**Another error that I found it worth to mention**, somehow I saw that each fold throws 0 weights in each feature. But it must not have been like this. 
1. I looked into the indices to be different in each fold. (checked, no issue observed here)
2. Data can be defective. Here we go, I noticed that X.iloc[wrong index], X.iloc[wrong index] The wrong index that was used before remained written. So no error was thrown but there is logic error here. Wrong data was evaluated here so I happened to face 0 importances.


### Another important note is why there is a mismatch btw feature importance and feature of X:

As the nature of one hot encoding, categorical values will be seperated such as [[0,0,1],[0,1,0]...]. It means that categorical features will multiple by the number of variety in columnns. That's why, there is a difference. When I try to plot the weights, it might be non-sense to see some breakdown of one hot encoded features. As the  feature of X is not competible with the feature of importance, the code will run into an issue! So I thought that that's enough to understand this :)

In [27]:
avg_importance

array([0.23189979, 0.14740156, 0.19743022, 0.08436801, 0.05656258,
       0.1577193 , 0.        , 0.016499  , 0.05221028, 0.04793846,
       0.        , 0.0014755 , 0.00159746, 0.00204485, 0.00082812,
       0.00202488])

In [28]:
len(avg_importance)

16

In [29]:
len(X.columns)

11